### ISSUE TITLE
arrow-parens: false positive for generic arrow function

### ISSUE THREAD

**Issue by JamesHenry** [typescript-eslint/typescript-eslint#14]
This issue was initially reported here: https://github.com/eslint/typescript-eslint-parser/issues/475

---

**What version of TypeScript are you using?**
2.8.3

**What version of `typescript-eslint-parser` are you using?**
15.0.0

**What code were you trying to parse?**

```ts
// Unexpected error: arrow-parens
const bar = <T>(t: T) => {};
// No error as expected
const bar2 = function <T>(t: T) {};
```

**Comment by jhoch** [comment]
@JamesHenry @armano2 I'm still seeing this on version `1.4.2`.  I have a file in my repo called `components/util/test.ts`:
```
const foo = <T>(bar: any): void => {
  // Do nothing
}
```
which produces the following when I run `eslint`:
```
jasonhoch@hoch1:~/code/newfront-app
15:34:48$ yarn eslint components/util/test.ts
yarn run v1.13.0
$ /Users/jasonhoch/code/newfront-app/node_modules/.bin/eslint components/util/test.ts

/Users/jasonhoch/code/newfront-app/components/util/test.ts
  1:7   warning  'foo' is assigned a value but never used                                             @typescript-eslint/no-unused-vars
  1:13  error    Expected parentheses around arrow function argument having a body with curly braces  arrow-parens
  1:17  warning  'bar' is defined but never used                                                      @typescript-eslint/no-unused-vars
  1:22  warning  Unexpected any. Specify a different type                                             @typescript-eslint/no-explicit-any
  3:2   error    Missing semicolon                                                                    semi

✖ 5 problems (2 errors, 3 warnings)
  2 errors and 0 warnings potentially fixable with the `--fix` option.

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```
(See the error at `1:13`).  Here is my `.eslintrc.js` file:

```
module.exports = {
    'extends': [
      'airbnb',
      'plugin:@typescript-eslint/recommended',
    ],
    'env': {
      'jest': true,
    },
    'overrides': [{
      'files': ['*.ts', '*.tsx'],
      'rules': {
        'indent': 'off', // Conflicts
        'semi': ['error', 'always'], // Conflicts
      },
    }],
    'parser': '@typescript-eslint/parser',
    'parserOptions': {
      'project': './tsconfig.json',
    },
    'plugins': [
      '@typescript-eslint',
      'import',
    ],
    'rules': {
      'arrow-body-style': ['off'],
      'jsx-a11y/anchor-is-valid': ['error', {
        'components': ['Link'],
        'aspects': ['invalidHref', 'preferButton'],
      }],
      'jsx-quotes': ['warn', 'prefer-single'],
      'react/jsx-filename-extension': ['warn', { 'extensions': ['.js', '.jsx', '.tsx'] }],
      'react/jsx-closing-tag-location': ['off'],
      'react/jsx-curly-brace-presence': ['error', { 'props': 'never', 'children': 'ignore' } ],
      'react/jsx-one-expression-per-line': ['off'],
      'react/prop-types': ['off'],
      'semi': ['warn', 'never'],
      '@typescript-eslint/explicit-function-return-type': ['warn', { 'allowExpressions': true }],
      '@typescript-eslint/indent': ['error', 2],
    },
    'settings': {
      'import/resolver': {
        'typescript': {},
      },
    },
}
```

and the lines in my `package.json`:

```
"devDependencies": {
    ...
    "@typescript-eslint/eslint-plugin": "^1.4.2",
    "@typescript-eslint/parser": "^1.4.2",
    ...
}
```

We've had a smooth experience moving over from TSLint, with this one glitch.  Looking at #176, it looks like you guys test for this case.  Any idea what the problem could be?  Let me know if I can help!

**Comment by jinasonlin** [comment]
I still get false positive.

#### `@typescript-eslint/eslint-plugin` version
1.7.1-alpha-28

#### rule
 "arrow-parens" : ["as-needed",  {  "requireForBlockBody": true }]

```javascript
// Expected parentheses around arrow function argument having a body with curly braces.eslint(arrow-parens)
const foo = <T>(t: T): void => { console.info(t); }

// the following is correct
const foo = (t) => { console.info(t); }
```

**Comment by bradzacher** [comment]
You have the missing piece of the puzzle!
The config that causes it to fail!

**Comment by tinymins** [comment]
I've the same problem, when will this problem get fixed and published?

**Comment by tinymins** [comment]
![image](https://user-images.githubusercontent.com/1808990/58616249-c9720c00-82ef-11e9-8a60-b987d6fb211b.png)

**Comment by oli-laban** [comment]
I'm also experiencing this with default `airbnb` `arrow-parens` config:

```
'arrow-parens': ['error', 'as-needed', {
  requireForBlockBody: true,
}],
```

```
// Expected parentheses around arrow function argument having a body with curly braces.
export const makeRequest = async <T>(
  method: RequestMethod,
  urlPath: string,
  body: {} = {},
  extraHeaders: RequestHeaders = {},
  throwGeneralError: boolean = true,
  suppressErrors: boolean = false,
): Promise<T | null> => {
  ...
};
```

```
"eslint": "^5.15.0",
"eslint-config-airbnb": "^17.1.0",
"eslint-import-resolver-alias": "^1.1.2",
"eslint-plugin-import": "^2.16.0",
"eslint-plugin-jsx-a11y": "^6.2.1",
"eslint-plugin-react": "^7.12.4",
"@typescript-eslint/eslint-plugin": "^1.9.0",
"@typescript-eslint/parser": "^1.9.0",
```

**Comment by aldarund** [comment]
Still issue. It happens when  { "requireForBlockBody": true } otherwise it don't happen.
Maybe it will get fixes one day :)

**Comment by mdjermanovic** [comment]
This is fixed in eslint v7.4.0
